src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -0,0 +1,472 @@
+            {
+                // If the property reference itself is a tainted data source
+                if (operation is IPropertyReferenceOperation propertyReferenceOperation
+                    && this.DataFlowAnalysisContext.SourceInfos.IsSourceProperty(propertyReferenceOperation.Property))
+                {
+                    return TaintedDataAbstractValue.CreateTainted(propertyReferenceOperation.Member, propertyReferenceOperation.Syntax, this.OwningSymbol);
+                }
+
+                if (AnalysisEntityFactory.TryCreate(operation, out AnalysisEntity analysisEntity))
+                {
+                    return this.CurrentAnalysisData.TryGetValue(analysisEntity, out TaintedDataAbstractValue value) ? value : defaultValue;
+                }
+
+                return defaultValue;
+            }
+
+            // So we can hook into constructor calls.
+            public override TaintedDataAbstractValue VisitObjectCreation(IObjectCreationOperation operation, object argument)
+            {
+                TaintedDataAbstractValue baseValue = base.VisitObjectCreation(operation, argument);
+                IEnumerable<IArgumentOperation> taintedArguments = operation.Arguments.Where(
+                    a => this.GetCachedAbstractValue(a).Kind == TaintedDataAbstractValueKind.Tainted
+                         && (a.Parameter.RefKind == RefKind.None
+                             || a.Parameter.RefKind == RefKind.Ref
+                             || a.Parameter.RefKind == RefKind.In));
+                if (taintedArguments.Any())
+                {
+                    ProcessTaintedDataEnteringInvocationOrCreation(operation.Constructor, taintedArguments, operation);
+                }
+
+                return baseValue;
+            }
+
+            public override TaintedDataAbstractValue VisitInvocation_NonLambdaOrDelegateOrLocalFunction(
+                IMethodSymbol method,
+                IOperation visitedInstance,
+                ImmutableArray<IArgumentOperation> visitedArguments,
+                bool invokedAsDelegate,
+                IOperation originalOperation,
+                TaintedDataAbstractValue defaultValue)
+            {
+                // Always invoke base visit.
+                TaintedDataAbstractValue baseVisit = base.VisitInvocation_NonLambdaOrDelegateOrLocalFunction(
+                    method, 
+                    visitedInstance,
+                    visitedArguments, 
+                    invokedAsDelegate, 
+                    originalOperation, 
+                    defaultValue);
+
+                IEnumerable<IArgumentOperation> taintedArguments = visitedArguments.Where(
+                    a => this.GetCachedAbstractValue(a).Kind == TaintedDataAbstractValueKind.Tainted
+                         && (a.Parameter.RefKind == RefKind.None
+                             || a.Parameter.RefKind == RefKind.Ref
+                             || a.Parameter.RefKind == RefKind.In));
```

Comment:
Nit: Seems like you can extract this code to a helper method, it looks similar to the one in object creation override.

Reply:
Possibly even the if statement below can be extracted into the helper.

---
In reply to: [234017052](https://github.com/dotnet/roslyn-analyzers/pull/1882#discussion_r234017052) [](ancestors = 234017052)